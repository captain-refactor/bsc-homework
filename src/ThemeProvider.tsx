import { createMuiTheme, MuiThemeProvider } from "@material-ui/core/styles";
import { PropsWithChildren } from "react";
import * as React from "react";

let theme = createMuiTheme({
  palette: {
    primary: {
      main: "#fc5f5f",
      contrastText: "#fefefe",
    },
    secondary: {
      main: "#067BC2",
      contrastText: "#fefefe",
    },
  },
});

export function ThemeProvider(props: PropsWithChildren<unknown>) {
  return <MuiThemeProvider theme={theme}>{props.children}</MuiThemeProvider>;
}
