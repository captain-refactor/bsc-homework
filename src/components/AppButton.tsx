import Button from "@material-ui/core/Button";
import React from "react";
import styled from "@material-ui/core/styles/styled";

export const AppButton = styled(Button)((theme) => ({
  background: "linear-gradient(45deg, #FE6B8B 30%, #FF8E53 90%)",
  color: "white",
  "&:focus": {
    outline: "solid 1px #FF8E53",
  },
}));
