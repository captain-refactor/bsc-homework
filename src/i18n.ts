import i18next from "i18next";
import { initReactI18next } from "react-i18next";

i18next.use(initReactI18next).init({
  lng: "en",
  resources: {
    en: {
      translation: {
        title: "Notes app",
        noteDetailHeading: "Note detail",
        notesListHeading: "Your notes",
        addNote: "Add note",
        newNoteHeading: "Create note",
        createNoteButton: "Create note",
        updateNoteButton: "Update note",
        deleteNoteButton: "Delete note",
      },
    },
    cs: {
      translation: {
        title: "Aplikace na poznámky",
        noteDetailHeading: "Detail poznámky",
        notesListHeading: "Vaše poznámky",
        addNote: "Přidat poznámku",
        newNoteHeading: "Vytvořit poznámku",
        createNoteButton: "Vytvořit poznámku",
        updateNoteButton: "Uložit poznámku",
        deleteNoteButton: "Smazat poznámku",
      },
    },
  },
});

export const I18N = i18next;
