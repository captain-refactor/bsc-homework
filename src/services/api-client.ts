export class ApiClient {
  apiBaseUrl = "https://private-anon-7762247fb2-note10.apiary-mock.com";

  getUrl(path: string) {
    return new URL(path, this.apiBaseUrl);
  }

  async fetch(path: string, options: RequestInit): Promise<Response> {
    let response = await fetch(this.getUrl(path).toString(), options);
    if (!response.ok) {
      throw new Error(response.statusText);
    }
    return response;
  }
}
