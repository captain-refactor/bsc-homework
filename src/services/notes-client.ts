import { ApiClient } from "./api-client";

export class NotesClient {
  constructor(private apiService: ApiClient) {}

  async fetchNotes(): Promise<ApiNote[]> {
    const response = await this.apiService.fetch("/notes", {});
    return await response.json();
  }

  async createNote(note: ApiNewNote): Promise<ApiNote> {
    const response = await this.apiService.fetch("/notes", {
      method: "POST",
      body: JSON.stringify(note),
    });
    return await response.json();
  }

  async fetchNote(noteId: number): Promise<ApiNote> {
    const response = await this.apiService.fetch(`/notes/${noteId}`, {});
    return await response.json();
  }

  async removeNote(noteId: number): Promise<void> {
    await this.apiService.fetch(`/notes/${noteId}`, {
      method: "DELETE",
    });
  }

  async updateNote(note: ApiUpdateNote): Promise<ApiNote> {
    const { id, ...body } = note;
    const response = await this.apiService.fetch(`/notes/${id}`, {
      method: "PUT",
      body: JSON.stringify(body),
    });
    return await response.json();
  }
}

interface ApiId {
  id: number;
}

interface ApiNoteData {
  title: string;
}

export interface ApiNote extends ApiId, ApiNoteData {}

export interface ApiNewNote extends ApiNoteData {}

export interface ApiUpdateNote extends ApiNoteData, ApiId {}
