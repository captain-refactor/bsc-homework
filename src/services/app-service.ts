import { NotesService } from "./notes-service";
import { ApiClient } from "./api-client";
import { NotesClient } from "./notes-client";

export class AppService {
  constructor(public notesService: NotesService) {}
}

export function createAppService() {
  let apiService = new ApiClient();
  return new AppService(new NotesService(new NotesClient(apiService)));
}
