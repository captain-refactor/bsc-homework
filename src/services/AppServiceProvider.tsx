import React, { createContext, PropsWithChildren, useContext } from "react";
import { AppService, createAppService } from "./app-service";

export const AppServiceContext = createContext<AppService>(null as any);

export class AppServiceProvider extends React.Component<Props> {
  state = {
    appService: createAppService(),
  };

  render() {
    return (
      <AppServiceContext.Provider
        value={this.props.appService || this.state.appService}
      >
        {this.props.children}
      </AppServiceContext.Provider>
    );
  }
}

type Props = PropsWithChildren<{
  appService?: AppService;
}>;

export const useAppService = () => useContext(AppServiceContext);
