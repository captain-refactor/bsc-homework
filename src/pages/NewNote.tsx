import { observer } from "mobx-react";
import Card from "@material-ui/core/Card";
import React, { useState } from "react";
import CardContent from "@material-ui/core/CardContent";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import { useAppService } from "../services/AppServiceProvider";
import { useTranslation } from "react-i18next";
import makeStyles from "@material-ui/core/styles/makeStyles";
import { ApiNewNote } from "../services/notes-client";
import TextField from "@material-ui/core/TextField";
import { AppButton } from "../components/AppButton";
import { useHistory } from "react-router";

export const NewNote = observer(function () {
  const app = useAppService();
  const { t } = useTranslation();
  const history = useHistory();
  const cls = useStyles();
  const [note, setNote] = useState<ApiNewNote>({ title: "" });

  return (
    <Card>
      <CardContent>
        <Grid container justify="center">
          <Grid item xs={12}>
            <Typography className={cls.title} variant="h2">
              {t("newNoteHeading")}
            </Typography>
          </Grid>
          <Grid item xs={10}>
            <TextField
              className={cls.textField}
              variant="outlined"
              multiline
              value={note.title}
              onChange={(event) => setNote({ title: event.target.value })}
            />
          </Grid>
          <Grid container item xs={12} direction="row-reverse">
            <AppButton
              onClick={() => {
                app.notesService.createNote(note);
                history.push("/notes");
              }}
            >
              {t("createNoteButton")}
            </AppButton>
          </Grid>
        </Grid>
      </CardContent>
    </Card>
  );
});

const useStyles = makeStyles((theme) => ({
  title: {
    fontSize: theme.spacing(4),
  },
  textField: {
    margin: theme.spacing(3, 0),
    width: "100%",
  },
}));
