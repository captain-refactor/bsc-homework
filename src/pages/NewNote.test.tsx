import { mount } from "enzyme";
import { ThemeProvider } from "../ThemeProvider";
import React from "react";
import { AppServiceProvider } from "../services/AppServiceProvider";
import { NewNote } from "./NewNote";
import { createAppService } from "../services/app-service";
import { ApiNewNote } from "../services/notes-client";
import { I18N } from "../i18n";
import { I18nextProvider } from "react-i18next";
import { act } from "react-dom/test-utils";
import { AppButton } from "../components/AppButton";
import TextField from "@material-ui/core/TextField";
import { MemoryRouter } from "react-router-dom";

describe("New note component", () => {
  it("should create new note", async () => {
    const appService = createAppService();

    let createdNote: ApiNewNote | undefined;

    appService.notesService.createNote = async (note) => {
      createdNote = note;
    };

    const wrapper = mount(
      <ThemeProvider>
        <I18nextProvider i18n={I18N}>
          <MemoryRouter>
            <AppServiceProvider appService={appService}>
              <NewNote />
            </AppServiceProvider>
          </MemoryRouter>
        </I18nextProvider>
      </ThemeProvider>
    );
    expect(wrapper.find(NewNote).html()).toMatchSnapshot();
    act(() => {
      wrapper
        .find(NewNote)
        .find(TextField)
        .find("textarea")
        .first()
        .simulate("change", { target: { value: "Test value" } });
    });
    act(() => {
      wrapper.find(NewNote).find(AppButton).simulate("click");
    });
    expect(createdNote?.title).toBe("Test value");
    wrapper.unmount();
  });
});
