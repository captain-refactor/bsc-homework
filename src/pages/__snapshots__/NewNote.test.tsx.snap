// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New note component should create new note 1`] = `"<div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiCardContent-root\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center\\"><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\"><h2 class=\\"MuiTypography-root makeStyles-title-1 MuiTypography-h2\\">Create note</h2></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10\\"><div class=\\"MuiFormControl-root MuiTextField-root makeStyles-textField-2\\"><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline\\"><textarea rows=\\"1\\" aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline\\" style=\\"height: 0px; overflow: hidden;\\"></textarea><textarea aria-hidden=\\"true\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline\\" readonly=\\"\\" tabindex=\\"-1\\" style=\\"visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;\\"></textarea><fieldset aria-hidden=\\"true\\" style=\\"padding-left: 8px;\\" class=\\"PrivateNotchedOutline-root-204 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legend-205\\" style=\\"width: 0.01px;\\"><span>â€‹</span></legend></fieldset></div></div></div><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-row-reverse MuiGrid-grid-xs-12\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text WithStyles(ForwardRef(Button))-root-208 WithStyles(ForwardRef(Button))-root-209\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Create note</span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></div></div>"`;
